### ISSUE TITLE
Performance drop on NVIDIA GPUs

### ISSUE THREAD

**Issue by tomdeakin** [LLNL/RAJA#144]
We've noted a performance drop in the RAJA version of [GPU-STREAM](https://github.com/UoB-HPC/GPU-STREAM) of varying degrees with RAJA running on a GTX 980 Ti GPU from NVIDIA. The issue also relates to driver versions, and different versions of RAJA; an older 5009f7bc75cb6fa92436f5cc291e1385e7ad2a86 and a relatively new 5b1ca99dbd76b59efea7f94fe5b70d559f9882d7.

Driver   | RAJA commit | Bandwidth (GB/s)
--------|--------------|-------------------
352.21  | 5009f7b          | 269
367.27 | 5b1ca99         | 197
352.63 | 5b1ca99        | ~250
370.28 | 5b1ca99        | ~250
370.28 | 5009f7b          | 269

The native CUDA port gives 269 GB/s, and this represents 80% theoretical peak.

**Comment by DavidPoliakoff** [comment]
Hey Tom,

Thanks for filing the issue. We've heard about this, it's good to get a reproducer. @jonesholger has been our main person looking into the driver side of things, I wanted to ping him, and I seem to remember @davidbeckingsale having some theories on why this is (I think we changed our default kernel launch pattern), I'll ping them to get them in the loop on this

**Comment by davidbeckingsale** [comment]
So, looking at the difference in the CUDA `forall` implementations:
- `5009f7b` is using the 1D grid, and just accessing the `.x` component when calculating which index to run.
- `5b1ca99` uses all 3 component when calculating which index to run, but all but the `.x` component are going to be 0, since they are still being launched on a 1D grid. Also, the device kernel which executes the lambda has changed - it now iterates such that each CUDA thread can handle multiple indices.

If you take a look at the attached PTX you can see the additional instructions caused by the 3D indexing.

[fast.ptx.txt](https://github.com/LLNL/RAJA/files/562952/fast.ptx.txt)
[slow.ptx.txt](https://github.com/LLNL/RAJA/files/562953/slow.ptx.txt)

**Comment by DavidPoliakoff** [comment]
It looks like we'll be able to back this out, thanks very much again for bringing it to our attention!

**Comment by tomdeakin** [comment]
@davidbeckingsale I tried it the branch out and it errors at code runtime:
```
CUDAassert: invalid device function /nfs/modules/raja/maxblocks-fix/include/RAJA/exec-cuda/forall_cuda.hxx 210
```

**Comment by davidbeckingsale** [comment]
@tomdeakin I think that Jason might have accidentally commited something else to that branch - can you try running with the branch `bugfix/cuda-1d-index`?

**Comment by tomdeakin** [comment]
The `bugfix/cuda-1d-index` branch has the same slightly low performance of 256 GB/s.

I made a faux pas with the `feature/burmark1/cuda-maxblocks-fix` branch and built with the wrong `-arch` flag :weary:. This branch also has the same slightly low performance of 256 GB/s.

**Comment by MrBurmark** [comment]
@tomdeakin Overhead introduced in changes made to the cuda reducers and cuda memutils probably also contributes to this change, especially with synchronous timing. I added guards that reduce the amount of checking done if no reducers are active in my `feature/burmark1/cuda-maxblocks-fix` branch.

**Comment by davidbeckingsale** [comment]
Thanks @MrBurmark 

@tomdeakin - what optimization flags are you compiling with?

**Comment by tomdeakin** [comment]
@davidbeckingsale `-arch=sm_52 -O3`

**Comment by tomdeakin** [comment]
@MrBurmark I ran on the GTX 980 Ti with the updated `feature/burmark1/cuda-maxblocks-fix` branch. It again ran at 256 GB/s.

**Comment by davidbeckingsale** [comment]
@tomdeakin I've pushed some more stuff to `bugfix/cuda-1d-index` that might address this. We are basically in the process of backing out features until we work out which one it was...

**Comment by tomdeakin** [comment]
It looks like there is now non-blocking behaviour on the `RAJA::forall`. Even adding in `cudaDeviceSynchronize()` calls still results in wrong answers.

**Comment by davidbeckingsale** [comment]
Give the branch in #155 a shot and let me know if it works for you. I'm observing the following results on a GTX 980 Ti with the branch in that PR:

```
Function    MBytes/sec  Min (sec)   Max         Average
Copy        262204.814  0.00205     0.00206     0.00205
Mul         262146.688  0.00205     0.00206     0.00205
Add         267665.472  0.00301     0.00302     0.00301
Triad       267643.232  0.00301     0.00302     0.00301
```

**Comment by tomdeakin** [comment]
I can replicate these numbers with the `bugfix/cuda-performance` at commit c5f69b3e62a3297

**Comment by davidbeckingsale** [comment]
Fixed by #155, @tomdeakin let us know if you notice anything else.
